### PR TITLE
Add Elo support

### LINKED ISSUE #57: Other's card type
Hello, what kind of 'Elo' and 'Aura' cards are not supported? Would you implement them please?

### LINKED ISSUE #60: Elo type support
Based on this one

https://github.com/moip/credit-card-validator/blob/master/src/main/java/br/com/moip/creditcard/EloCreditCard.java

### LINKED ISSUE #65: [Feature] CardType from Brazil: Elo
### General information

* Library version: 6.0.0
* Browser and OS: Chrome 63.0.3239.132 (64-bit) on MacOS 10.13.2 

### Issue description

I'm from brazil, and here, we have a credit card type called "Elo", i want that type of card be accepted in this library ... so, i will send a pull request to this.

### LINKED ISSUE #66: [Official] Elo credit card documentation
Hello, this issue is just to request this implementation and check if this is all that you needed to push this feature from backlog.

Official Website: https://dev.elo.com.br/documentacao/tabela-de-bins

This website doesn't a translate to english.

### PR DESCRIPTION
**Pull request by crookedneighbor**
Adds Elo support to credit card type.

Elo bin ranges collide with Visa, Maestro, Discover and UnionPay.

We made the decision for Visa to have it detect as Visa unless we are certain that it is an Elo card.

The same goes for some, but not all, bin ranges for Maestro, Discover and UnionPay.

closes #57
closes #60 
closes #65
closes #66

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -148,6 +226,120 @@ types[MAESTRO] = {
...
+    '|' +
+    '4514|45141|451416' +
+    '|' +
+    '457|457[36]|45739|45763|457393|45763[12]' +
+
+    '|' +
+
+    '50|50[69]' +
+    '|' +
+    '506[6-7]|50669|5067[0-7]|5067[0-6][0-9]|50677[0-8]' +
```
Comment (crookedneighbor) [inline index.js]: I know `[0-9]` is equivalent to `\d`, but I felt like having the range made this (slightly) easier to read.

**Review by jackellenberger — APPROVED** [review]
this is gross and i love it

File: index.js [hunk 2]
```diff
@@ -148,6 +226,120 @@ types[MAESTRO] = {
...
+    '627780' +
+
+    '|' +
+
+    '636(297|368)' +
+
+    '|' +
+
+    // Dual Branded with Visa
+    '4(0117[89]|38935|5763[12])' +
```
Comment (chunhuiho) [inline index.js]: For the bin started with 4, we found thats not a dual branded card. Those are old BINs that were utilized before ELO had established the partnership with Discover. So we just remap these bins to Elo only card.

Comment (crookedneighbor) [reply index.js]: Okay, I'll update the comment.

File: test/index.js [hunk 3]
```diff
@@ -176,6 +191,49 @@ describe('creditCardType', function () {
...
+    var tests = [
+      ['43', 'visa'],
+      ['431', 'visa'],
+      ['4312', 'visa'],
+      ['43127', 'visa'],
+      ['45141', 'visa'],
+      ['45739', 'visa'],
+      ['40117', 'visa'],
+      ['43893', 'visa'],
+      ['45763', 'visa'],
```
Comment (chunhuiho) [inline test/index.js]: Same as above, we found remap those bins started with 4 to just be Elo only card.

Comment (crookedneighbor) [reply test/index.js]: I think we'll adjust this to show `'visa', 'elo'` until you have put in the 6th digit. Then it will flip to just visa or just elo.

Comment (chunhuiho) [reply test/index.js]: hmm...unfortunately I don't have it either, I guess Visa own most of the bin range started with 4? Is it reasonable to recognize as visa until we found that's a elo?

Comment (crookedneighbor) [reply test/index.js]: That is what we are currently doing. I think for this library, we'll show both until the 6th digit, but for Hosted Fields, we'll just pick the first one (Visa) until the 6th digit, and then it'll either flip to Elo or remain visa.

Comment (chunhuiho) [reply test/index.js]: that sounds good to me. we can hold it until user put in 6 digits

**Comment by crookedneighbor** [comment]
We made the adjustments talked about here: https://github.com/braintree/credit-card-type/pull/76#discussion_r206633629

When used within hosted fields, we take the first match anyway, so even if more than 1 card type is shown, it'll still show the first match and Elo has lower priority than the rest.

File: index.js [hunk 4]
```diff
@@ -139,7 +198,26 @@ types[MAESTRO] = {
...
+        '|' +
+        '9[0-9][0-9][0-9]' +
+      ')\\d{4}' +
+    '|' +
+    '67' +
+    '|' +
+    '63' +
+      '(?!' + // More Elo card ranges
+        '6297|6368' +
+      ')\d{4}' +
```
Comment (jackellenberger) [inline index.js]: \d -> \\d

**Review by braintreeps — APPROVED** [review]

**Review by chunhuiho — APPROVED** [review]

**Review by quinnjn — APPROVED** [review]
Those are some nasty looking regular expressions!

As card BIN identification gets more complex should we think about refactoring this library in the future to take a `function` to process the pattern?

**Comment by crookedneighbor** [comment]
@jackellenberger and I were thinking the same thing.

**Pull request merged**